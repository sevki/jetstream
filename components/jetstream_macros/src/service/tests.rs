#![cfg(test)]

use super::service_impl;
use core::panic;
use syn::parse_quote;

fn run_test_with_filters<F>(test_fn: F)
where
    F: FnOnce() + panic::UnwindSafe,
{
    let filters = vec![
        // Filter for protocol version strings
        (
            r"dev\.branch\.jetstream\.proto/\w+/\d+\.\d+\.\d+-[a-f0-9]{8}",
            "dev.branch.jetstream.proto/NAME/VERSION-HASH",
        ),
        // Filter for digest strings
        (r"[a-f0-9]{64}", "DIGEST_HASH"),
    ];

    insta::with_settings!({
        filters => filters,
    }, {
        test_fn();
    });
}

#[test]
fn test_simple_service() {
    let input: syn::ItemTrait = parse_quote! {
        pub trait Echo {
            async fn ping(&self) -> Result<(), std::io::Error>;
        }
    };
    let output = service_impl(input, false, false);
    let syntax_tree: syn::File = syn::parse2(output).unwrap();
    let output_str = prettyplease::unparse(&syntax_tree);
    run_test_with_filters(|| {
        insta::assert_snapshot!(output_str);
    })
}

#[test]
fn test_service_with_args() {
    let input: syn::ItemTrait = parse_quote! {
        pub trait Echo {
            async fn ping(&self, message: String) -> Result<String, std::io::Error>;
        }
    };
    let output = service_impl(input, false, false);
    let syntax_tree: syn::File = syn::parse2(output).unwrap();
    let output_str = prettyplease::unparse(&syntax_tree);
    run_test_with_filters(|| {
        insta::assert_snapshot!(output_str);
    })
}

#[test]
fn test_async_trait_service_with_args() {
    let input: syn::ItemTrait = parse_quote! {
        pub trait Echo {
            async fn ping(&mut self, message: String) -> Result<String, std::io::Error>;
        }
    };
    let output = service_impl(input, true, false);
    let syntax_tree: syn::File = syn::parse2(output).unwrap();
    let output_str = prettyplease::unparse(&syntax_tree);
    run_test_with_filters(|| {
        insta::assert_snapshot!(output_str);
    })
}
