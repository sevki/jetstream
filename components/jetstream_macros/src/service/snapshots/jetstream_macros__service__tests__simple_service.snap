---
source: components/jetstream_macros/src/service/tests.rs
expression: output_str
---
pub mod echo_protocol {
    use jetstream::prelude::*;
    use std::io::{self, Read, Write};
    use std::mem;
    use super::Echo;
    const MESSAGE_ID_START: u8 = 101;
    pub const PROTOCOL_VERSION: &str = "dev.branch.jetstream.proto/NAME/VERSION-HASH";
    const DIGEST: &str = "DIGEST_HASH";
    pub const TPING: u8 = MESSAGE_ID_START + 0u8;
    pub const RPING: u8 = MESSAGE_ID_START + 0u8 + 1;
    #[allow(non_camel_case_types)]
    #[derive(Debug, JetStreamWireFormat)]
    pub struct Tping {}
    #[allow(non_camel_case_types)]
    #[derive(Debug, JetStreamWireFormat)]
    pub struct Rping(pub ());
    #[derive(Debug)]
    #[repr(u8)]
    pub enum Tmessage {
        Ping(Tping) = TPING,
    }
    impl Framer for Tmessage {
        fn byte_size(&self) -> u32 {
            match &self {
                Tmessage::Ping(msg) => msg.byte_size(),
            }
        }
        fn message_type(&self) -> u8 {
            unsafe { *<*const _>::from(self).cast::<u8>() }
        }
        fn encode<W: Write>(&self, writer: &mut W) -> io::Result<()> {
            match &self {
                Tmessage::Ping(msg) => msg.encode(writer)?,
            }
            Ok(())
        }
        fn decode<R: Read>(reader: &mut R, ty: u8) -> io::Result<Tmessage> {
            match ty {
                TPING => Ok(Tmessage::Ping(WireFormat::decode(reader)?)),
                _ => {
                    Err(
                        std::io::Error::new(
                            std::io::ErrorKind::InvalidData,
                            format!("unknown message type: {}", ty),
                        ),
                    )
                }
            }
        }
    }
    #[derive(Debug)]
    #[repr(u8)]
    pub enum Rmessage {
        Ping(Rping) = RPING,
    }
    impl Framer for Rmessage {
        fn byte_size(&self) -> u32 {
            match &self {
                Rmessage::Ping(msg) => msg.byte_size(),
            }
        }
        fn message_type(&self) -> u8 {
            unsafe { *<*const _>::from(self).cast::<u8>() }
        }
        fn encode<W: Write>(&self, writer: &mut W) -> io::Result<()> {
            match &self {
                Rmessage::Ping(msg) => msg.encode(writer)?,
            }
            Ok(())
        }
        fn decode<R: Read>(reader: &mut R, ty: u8) -> io::Result<Rmessage> {
            match ty {
                RPING => Ok(Rmessage::Ping(WireFormat::decode(reader)?)),
                _ => {
                    Err(
                        std::io::Error::new(
                            std::io::ErrorKind::InvalidData,
                            format!("unknown message type: {}", ty),
                        ),
                    )
                }
            }
        }
    }
    #[derive(Clone, Debug)]
    pub struct EchoService<T: Echo> {
        pub inner: T,
    }
    impl<T> Protocol for EchoService<T>
    where
        T: Echo + Send + Sync + Sized,
    {
        type Request = Tmessage;
        type Response = Rmessage;
        type Error = Error;
        const VERSION: &'static str = PROTOCOL_VERSION;
        fn rpc(
            &mut self,
            ctx: Context,
            frame: Frame<<Self as Protocol>::Request>,
        ) -> impl ::core::future::Future<
            Output = Result<Frame<<Self as Protocol>::Response>, Self::Error>,
        > + Send + Sync {
            Box::pin(async move {
                let req: <Self as Protocol>::Request = frame.msg;
                let res: Result<<Self as Protocol>::Response, Self::Error> = match req {
                    Tmessage::Ping(msg) => {
                        let result = self.ping().await?;
                        let ret = Rping(result);
                        Ok(Rmessage::Ping(ret))
                    }
                };
                let rframe: Frame<<Self as Protocol>::Response> = Frame::from((
                    frame.tag,
                    res?,
                ));
                Ok(rframe)
            })
        }
    }
    impl<T> Echo for EchoService<T>
    where
        T: Echo + Send + Sync + Sized,
    {
        async fn ping(&self) -> Result<(), std::io::Error> {
            self.inner.ping().await
        }
    }
    pub struct EchoChannel<'a> {
        pub inner: Box<&'a mut dyn ClientTransport<Self>>,
    }
    impl<'a> Protocol for EchoChannel<'a> {
        type Request = Tmessage;
        type Response = Rmessage;
        type Error = Error;
        const VERSION: &'static str = PROTOCOL_VERSION;
        fn rpc(
            &mut self,
            _ctx: Context,
            frame: Frame<<Self as Protocol>::Request>,
        ) -> impl ::core::future::Future<
            Output = Result<Frame<<Self as Protocol>::Response>, Self::Error>,
        > + Send + Sync {
            use futures::{SinkExt, StreamExt};
            Box::pin(async move {
                self.inner.send(frame).await?;
                let frame = self.inner.next().await.unwrap()?;
                Ok(frame)
            })
        }
    }
    lazy_static::lazy_static! {
        static ref ECHO_TAG : std::sync::atomic::AtomicU16 =
        std::sync::atomic::AtomicU16::new(0);
    }
    impl<'a> Echo for EchoChannel<'a> {
        async fn ping(&mut self) -> Result<(), std::io::Error> {
            let tag = ECHO_TAG.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
            let req = Tmessage::Ping(Tping {});
            let tframe = Frame::from((tag, req));
            let context = Context::default();
            let rframe = self.rpc(context, tframe).await?;
            let rmsg = rframe.msg;
            match rmsg {
                Rmessage::Ping(msg) => Ok(msg.0),
                _ => Err(Error::InvalidResponse),
            }
        }
    }
}
#[jetstream::prelude::make(Send+Sync)]
pub trait Echo {
    async fn ping(&self) -> Result<(), std::io::Error>;
}
